apps/passport-server/src/services/generic-issuance/pipelines/PODPipeline/utils/finalizeAtom.ts

```diff
@@ -0,0 +1,48 @@
+import {
+  PODPipelineOutput,
+  VerifiedCredentialWithEmail
+} from "@pcd/passport-interface";
+import { PODEntries } from "@pcd/pod";
+import { PODAtom } from "../PODPipeline";
+
+/**
+ * Atoms are built during Pipeline load, using values from the pipeline
+ * input. Constant "configured" values and values extracted from the user's
+ * credential are not included at that time.
+ *
+ * This function takes the atom, the pipeline output, and the user's credential,
+ * and returns the final POD entries before signing.
+ *
+ * @param atom The atom containing POD entries
+ * @param output The pipeline output configuration
+ * @param credential The user's credential
+ * @returns The finalized POD entries
+ */
+export function finalizeAtom(
+  atom: PODAtom,
+  output: PODPipelineOutput,
+  credential: VerifiedCredentialWithEmail
+): PODEntries {
+  const newEntries: PODEntries = {};
+
+  for (const [name, entry] of Object.entries(output.entries)) {
+    if (entry.source.type === "configured") {
+      newEntries[name] = {
+        type: "string",
+        value: entry.source.value
+      };
+    } else if (entry.source.type === "credentialSemaphoreID") {
+      newEntries[name] = {
+        type: "cryptographic",
+        value: BigInt(credential.semaphoreId)
+      };
+    } else if (entry.source.type === "credentialEmail") {
+      newEntries[name] = {
+        type: "string",
+        value: credential.email
+      };
+    }
+  }
+
+  return { ...atom.entries, ...newEntries };
```

Comment:
As I mentioned in atomize, it might be desirable to "canonicalize" the entries here, by returning them as PODContent, or else round-tripping via PODContent.fromEtnries(...).asEntries().

Reply:
FWIW now that you've raised the performance concern I withdraw this suggestion at least in literal form.
The only actual change that "canonicalization" does to entries is putting them in order by name, so that's something you might consider doing before using them as input to a hash, so as to ensure stability.

Reply:
`PODContent` can sort the entries without producing a content ID, so that seems like a reasonable step that avoids the need to build the Merkle tree but does ensure that the PODEntries are predictably ordered.